pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java

```diff
@@ -290,18 +298,183 @@ void markResetStream() throws Exception {
+      InputStream bis = blob.getInputStream();
+      assertEquals(0, bis.read());
+      assertEquals(1024L, bis.skip(1024));
+      assertEquals(1, bis.read());
+      assertEquals(64 * 1024L, bis.skip(64 * 1024));
+      assertEquals(65, bis.read());
+    }
+  }
+
+  @Test
+  void skipBackwards() throws Exception {
+    LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+    long loid = createMediumLargeObject();
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(0, bis.read());
+      assertEquals(1024L, bis.skip(1024));
+      assertEquals(1, bis.read());
+      assertEquals(0L, bis.skip(-1024));
+      assertEquals(1, bis.read());
+    }
+  }
+
+  @Test
+  void skipToEnd() throws Exception {
+    LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+    long loid = createMediumLargeObject();
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(96 * 1024, bis.skip(96 * 1024));
+      assertEquals(-1, bis.read());
+    }
+  }
+
+  @Test
+  void skipPastEnd() throws Exception {
+    LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+    long loid = createMediumLargeObject();
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(1024 * 1024, bis.skip(1024 * 1024));
+      assertEquals(-1, bis.read());
+      assertEquals(1024, bis.skip(1024));
+      assertEquals(-1, bis.read());
+    }
+  }
+
+  @Test
+  void skipPastMaxAfter9_3() throws Exception {
+    assumeTrue(TestUtil.haveMinimumServerVersion(con, ServerVersion.v9_3));
+
+    LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+    long loid = createMediumLargeObject();
+
+    AtomicBoolean completedTest = new AtomicBoolean(false);
+    PSQLException ex = assertThrows(PSQLException.class, () -> {
+      try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
         InputStream bis = blob.getInputStream();
+        assertEquals(0, bis.read());
+        assertThrows(IOException.class, () -> bis.skip(Long.MAX_VALUE / 2));
+        assertEquals(0, bis.read());
+        assertThrows(IOException.class, () -> {
+          while (bis.read() != -1) {
+            // consume stream
+          }
+        });
+        assertThrows(IOException.class, bis::close);
+        completedTest.set(true);
+      }
+    });
 
-        assertEquals('<', bis.read());
-        bis.mark(4);
-        assertEquals('?', bis.read());
-        assertEquals('x', bis.read());
-        assertEquals('m', bis.read());
-        assertEquals('l', bis.read());
-        bis.reset();
-        assertEquals('?', bis.read());
+    assertEquals(PSQLState.IN_FAILED_SQL_TRANSACTION.getState(), ex.getSQLState());
+    assertTrue(completedTest.get());
```

Comment:
Can you please clarify how this can be `true` along with `PSQLException ex = assertThrows(PSQLException.class, () -> {` above?

Does it mean you expect an error out of `blob.close()`?
If that is the case, I would structure the test explicitly, without try-with-resources, otherwise it is hard to understand what is actually tested here and how it is supposed to work

Reply:
Yes, it throws on close. I'll change it as suggested.